daffodil-core/src/main/scala/org/apache/daffodil/core/dpath/Expression.scala

```diff
@@ -969,16 +989,27 @@ sealed abstract class StepExpression(val step: String, val pred: Option[Predicat
   }
 
   final lazy val isArray: Boolean = {
-    val (arrays, scalars) = stepElements.partition { _.isArray }
-    (arrays.length, scalars.length) match {
+    checkAmbiguousPath
+    stepElements.exists(_.isArray)
+  }
+
+  final lazy val isOptional: Boolean = {
+    checkAmbiguousPath
+    stepElements.exists(_.isOptional)
+  }
```

Comment:
These feel potentially confusing. I feel like our definition is usually an element is either an array OR an optional, but never both. But with step expressions, because of ambiguities, it's possible for isArray and isOptional to both be true. That seems reasonable considering the way dpath works and supports this ambiguity, but it might be worth adding a comment.

Reply:
Agreed.

Reply:
I am not sure we should allow this kind of polymorphism. We don't allow array and scalar polymorphism, nor optional and scalar polymorphism. I think allowing array and optional polymorphism is unwise in this context. 

If you have a path. There should be only one way it needs to be compiled - the path steps should be exactly the same on all paths or it should be an error. That's certainly the way to avoid what is otherwise a quite subtle situation that I would claim, is likely a mistake.  Arrays and optionals have different downward steps (DownArrayOccurrence vs. DownElement).

Reply:
The isOptional function is now renamed to isArrayorOptional to be clearer about its purpose.